File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3080_mark-elements-on-array-by-performing-queries/solutions/python3/solution.py
class Solution:
    def unmarkedSumArray(self, nums: list[int], queries: list[list[int]]) -> list[int]:
        from sortedcontainers import SortedList
        
        n = len(nums)
        marked = [False] * n
        answer = []
        unmarked_values = SortedList()
        
        # Initialize the unmarked values sorted list
        for value in nums:
            unmarked_values.add(value)
        
        for index, k in queries:
            # Mark the element at the given index if it's not already marked
            if not marked[index]:
                marked[index] = True
                unmarked_values.remove(nums[index])
            
            # Mark the k smallest unmarked elements
            for _ in range(min(k, len(unmarked_values))):
                smallest_unmarked = unmarked_values.pop(0)
                smallest_unmarked_index = nums.index(smallest_unmarked)
                marked[smallest_unmarked_index] = True
            
            # Sum of unmarked elements after the current query
            answer.append(sum(unmarked_values))
        
        return answer

File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3080_mark-elements-on-array-by-performing-queries/solutions/cpp/solution.cpp
#include <vector>
#include <set>

using namespace std;

class Solution {
public:
    vector<long long> unmarkedSumArray(vector<int>& nums, vector<vector<int>>& queries) {
        int n = nums.size();
        int m = queries.size();
        vector<bool> marked(n, false);
        multiset<pair<int, int>> unmarked_elements;
        long long sum_unmarked = 0;

        // Initialize all elements as unmarked with their value and index
        for (int i = 0; i < n; i++) {
            unmarked_elements.insert({nums[i], i});
            sum_unmarked += nums[i];
        }

        vector<long long> result;

        for (auto& query : queries) {
            int index = query[0];
            int k = query[1];

            // First, mark the element at the given index if it is not already marked
            if (!marked[index]) {
                auto it = unmarked_elements.find({nums[index], index});
                if (it != unmarked_elements.end()) {
                    sum_unmarked -= nums[index];
                    unmarked_elements.erase(it);
                }
                marked[index] = true;
            }

            // Mark k smallest unmarked elements
            auto it = unmarked_elements.begin();
            while (k > 0 && it != unmarked_elements.end()) {
                auto [val, idx] = *it;
                sum_unmarked -= val;
                it = unmarked_elements.erase(it);
                marked[idx] = true;
                k--;
            }

            // Store the current sum of unmarked elements
            result.push_back(sum_unmarked);
        }

        return result;
    }
};
